pkg/restore-util/client.go

```diff
@@ -0,0 +1,167 @@
+)
+
+// Client is a external client used by RegionSplitter.
+type Client interface {
+	// GetStore gets a store by a store id.
+	GetStore(ctx context.Context, storeID uint64) (*metapb.Store, error)
+	// GetRegion gets a region which includes a specified key.
+	GetRegion(ctx context.Context, key []byte) (*RegionInfo, error)
+	// GetRegionByID gets a region by a region id.
+	GetRegionByID(ctx context.Context, regionID uint64) (*RegionInfo, error)
+	// SplitRegion splits a region from a key, if key is not included in the region, it will return nil.
+	SplitRegion(ctx context.Context, regionInfo *RegionInfo, key []byte) (*RegionInfo, error)
+	// ScatterRegion scatters a specified region.
+	ScatterRegion(ctx context.Context, regionInfo *RegionInfo) error
+	// GetOperator gets the status of operator of the specified region.
+	GetOperator(ctx context.Context, regionID uint64) (*pdpb.GetOperatorResponse, error)
+	// ScanRegion gets a list of regions, starts from the region that contains key.
+	// Limit limits the maximum number of regions returned.
+	ScanRegions(ctx context.Context, key []byte, limit int) ([]*RegionInfo, error)
+}
+
+// pdClient is a wrapper of pd client, can be used by RegionSplitter.
+type pdClient struct {
+	mu         sync.Mutex
+	client     pd.Client
+	storeCache map[uint64]*metapb.Store
+}
+
+// NewClient returns a client used by RegionSplitter.
+func NewClient(client pd.Client) (Client, error) {
+	return &pdClient{
+		client:     client,
+		storeCache: make(map[uint64]*metapb.Store),
+	}, nil
+}
+
+func (c *pdClient) GetStore(ctx context.Context, storeID uint64) (*metapb.Store, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	store, ok := c.storeCache[storeID]
+	if ok {
+		return store, nil
+	}
+	store, err := c.client.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, err
+	}
+	c.storeCache[storeID] = store
+	return store, nil
+
+}
+
+func (c *pdClient) GetRegion(ctx context.Context, key []byte) (*RegionInfo, error) {
+	region, leader, err := c.client.GetRegion(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	return &RegionInfo{
+		Region: region,
+		Leader: leader,
+	}, nil
+}
+
+func (c *pdClient) GetRegionByID(ctx context.Context, regionID uint64) (*RegionInfo, error) {
+	region, leader, err := c.client.GetRegionByID(ctx, regionID)
+	if err != nil {
+		return nil, err
+	}
+	return &RegionInfo{
+		Region: region,
+		Leader: leader,
+	}, nil
+}
+
+func (c *pdClient) SplitRegion(ctx context.Context, regionInfo *RegionInfo, key []byte) (*RegionInfo, error) {
+	var peer *metapb.Peer
+	if regionInfo.Leader != nil {
+		peer = regionInfo.Leader
+	} else {
+		peer = regionInfo.Region.Peers[0]
+	}
+	reqCtx := &kvrpcpb.Context{
+		RegionId:    regionInfo.Region.Id,
+		RegionEpoch: regionInfo.Region.RegionEpoch,
+		Peer:        peer,
+	}
+	storeID := peer.GetStoreId()
+	store, err := c.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, err
+	}
+	req := &kvrpcpb.SplitRegionRequest{
+		Context:   reqCtx,
```

Comment:
Since `reqCtx` is only used once I prefer inlining that struct literal here 🙂 (or maybe even inline `req` into line 112).

```suggestion
		Context:   &kvrpcpb.Context{
			RegionId:    regionInfo.Region.Id,
			RegionEpoch: regionInfo.Region.RegionEpoch,
			Peer:        peer,
		},
```